### PR TITLE
Recursive flatten color and backgroundImage

### PR DESCRIPTION
**Pull request by lbennett-stacki**
Recursively `flattenColorPalette` and use to flatten `backgroundImage` utility declarations.

Docs: tailwindlabs/tailwindcss.com#562

### CODE REVIEW

File: __tests__/flattenColorPalette.test.js [hunk 1]
```diff
@@ -43,3 +43,43 @@ test('it flattens nested color objects', () => {
     'blue-3': 'rgb(0,0,100)',
   })
 })
+
+test('it recusively flattens nested color objects', () => {
```
Comment (LukeeeeBennett) [inline __tests__/flattenColorPalette.test.js:47]:
```suggestion
test('it recursively flattens nested color objects', () => {
```

File: __tests__/plugins/backgroundImage.test.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        '.bg-hero': {
+          'background-image': 'url("./hero.png")',
+        },
+      },
+      [],
+    ],
+  ])
+})
+
+test('it creates background-image utilities recusively', () => {
```
Comment (LukeeeeBennett) [inline __tests__/plugins/backgroundImage.test.js:58]:
```suggestion
test('it creates background-image utilities recursively', () => {
```

**Comment by LukeeeeBennett** [comment]
I've opened a docs PR at https://github.com/tailwindlabs/tailwindcss.com/pull/562

**Comment by snack-0verflow** [comment]
@LukeeeeBennett This feature looks really neat! Looking forward to using nested colors

**Comment by adamwathan** [comment]
Hey thanks! Can you please write up a clear and detailed description of what this does and include it in the PR description? I keep coming to this and when I see that I have to read through all the code to figure out what this PR is for I just keep pushing it off to work on things I can get through more quickly.

**Comment by adamwathan** [comment]
Never mind read the tests and understand it. I am working on making the whole config behave more consistently so I think we need to think through if this should apply to every single utility as just standard behavior or if it should really be unique to colors and background images.

I'm also exploring supporting deep merging of colors when using "extend" and have to think through how this might interact with that idea 🤔

**Comment by LukeeeeBennett** [comment]
> Hey thanks! Can you please write up a clear and detailed description of what this does and include it in the PR description? I keep coming to this and when I see that I have to read through all the code to figure out what this PR is for I just keep pushing it off to work on things I can get through more quickly.

Sorry @adamwathan, very lazy of me.

This allows for _infinitly_ deeply nested color palettes using recursion. A good example to check is on the docs PR, you can see a preview here https://tailwindcss-com-git-recursive-flatten-color-bg-image.tailwindlabs.vercel.app/docs/customizing-colors#nested-object-syntax. See how we now have a dot path for `theme.colors.indigo.dark.default` which outputs `*-indigo-dark-default` classes, which was not previously possible.

I then also use the new deep flattenColorPalette to parse deeply nested backgroundImage declarations. You can see the docs preview here https://tailwindcss-com-git-recursive-flatten-color-bg-image.tailwindlabs.vercel.app/docs/background-image#background-images where we have the now parsable dot path `theme.extend.backgroundImage.brand.logo.default` which is used to output `bg-brand-logo-default` classes.

This is useful to me building a product that has a slightly more complex design system that utilizes hierarchy to organize/define its palettes. This is used to theme the same UI product dynamically in multiple ways whilst providing styling configuration to some end users.

One really quick example is dark/light theming.

Instead of

```
colors: {
    'primary-dark': { default: '...' },
    // ... *-dark
    'primary-light': { default: '...' },
   /// ... *-light
},
```

we can

```
colors: {
    dark: {
        primary: { default: '...' }
        // ...
    },
    light: {
        primary: { light: '...' }
        // ...
    },
}

**Comment by LukeeeeBennett** [comment]
> Never mind read the tests and understand it. I am working on making the whole config behave more consistently so I think we need to think through if this should apply to every single utility as just standard behavior or if it should really be unique to colors and background images.
> 
> I'm also exploring supporting deep merging of colors when using "extend" and have to think through how this might interact with that idea 🤔

Agreed. Ideally, this is applied across all plugins that read the config. As suggested this seemed a little overwhelming for me so I decided to tackle the cases most important to me. Happy to help move this across plugins to make sure unit tests are available, and then we/you can look at a more holistic approach where `flattenColorPalette` is part of the parsing path for every possible config item.

**Comment by adamwathan** [comment]
I realized that a PR for this (aside from the background image stuff) already exists here:

https://github.com/tailwindlabs/tailwindcss/pull/2148

So since that PR was first I'm going to close this one sorry 😞  I think for now I would like to contain the scope of this recursive flattening stuff to just colors, and we can explore adding it elsewhere if there's enough demand for it.

**Comment by LukeeeeBennett** [comment]
Sure, seems a little odd to me and have the coverage so will just use my fork from now, thanks!

Should close the docs PR too.

**Pull request closed without merging**